.github/workflows/ci.yml

```diff
@@ -102,7 +103,7 @@ jobs:
     name: target (${{ matrix.pretty }},${{ matrix.os }})
     runs-on: ${{ matrix.os }}
     needs: [shellcheck, test, generate-matrix, check]
-    if: github.event_name == 'push'
+    if: ${{ github.event_name == 'push' && needs.generate-matrix.outputs.matrix != '{}' && needs.generate-matrix.outputs.matrix != '[]' && needs.generate-matrix.outputs.matrix != '' }}
```

Comment:
I think this should take into account if we are in a try or not, and only allow empty on a try

Reply:
I have no idea how to do that in an `if` statement (without a run branch).

Reply:
```yml
if: |
  github.event_name == 'push' && 
  (needs.generate-matrix.outputs.matrix != '{}' && needs.generate-matrix.outputs.matrix != '[]' && needs.generate-matrix.outputs.matrix != '' && (github.ref == 'refs/heads/trying' || true))
```

seems correct?

if push, and matrix not empty or not in trying

Reply:
We could probably have the preceeding step store an envvar and check `env.MATRIX_IS_EMPTY` or something? Since `|| true` will always be true. Would make the statement easier then too.

Reply:
yes, or even simpler, have generate-matrix output an object `{"empty": true}` on empty when not in try

Reply:
even even simpler, have generate-matrix fail when it's empty on a non-trying branch

Reply:
```
diff --git a/xtask/src/ci.rs b/xtask/src/ci.rs
index 012be66..ec348f5 100644
--- a/xtask/src/ci.rs
+++ b/xtask/src/ci.rs
@@ -27,6 +27,9 @@ pub enum CiJob {
         ref_name: String,
     },
     TargetMatrix {
+        // main, v0.1.0
+        #[clap(long, env = "GITHUB_REF_NAME")]
+        ref_name: String,
         #[clap(long, env = "COMMIT_MESSAGE")]
         message: String,
         #[clap(long, env = "COMMIT_AUTHOR")]
@@ -114,8 +117,12 @@ pub fn ci(args: CiJob, metadata: CargoMetadata) -> cross::Result<()> {
                 }
             }
         }
-        CiJob::TargetMatrix { message, author } => {
-            target_matrix::run(message, author)?;
+        CiJob::TargetMatrix {
+            ref_name,
+            message,
+            author,
+        } => {
+            target_matrix::run(ref_name, message, author)?;
         }
     }
     Ok(())
diff --git a/xtask/src/ci/target_matrix.rs b/xtask/src/ci/target_matrix.rs
index e8c87dd..6db8446 100644
--- a/xtask/src/ci/target_matrix.rs
+++ b/xtask/src/ci/target_matrix.rs
@@ -4,7 +4,11 @@ use serde::Serialize;
 
 use crate::util::{get_matrix, gha_output, gha_print, CiTarget, ImageTarget};
 
-pub(crate) fn run(message: String, author: String) -> Result<(), color_eyre::Report> {
+pub(crate) fn run(
+    ref_name: String,
+    message: String,
+    author: String,
+) -> Result<(), color_eyre::Report> {
     let mut matrix: Vec<CiTarget> = get_matrix().clone();
     if author == "bors[bot]" && message.starts_with("Try #") {
         if let Some((_, args)) = message.split_once(": ") {
@@ -32,7 +36,11 @@ pub(crate) fn run(message: String, author: String) -> Result<(), color_eyre::Rep
             std: target.std.map(|b| b as u8),
         })
         .collect::<Vec<_>>();
-    let json = serde_json::to_string(&matrix)?;
+    let json = if !matrix.is_empty() || ref_name == "trying" {
+        serde_json::to_string(&matrix)?
+    } else {
+        eyre::bail!("matrix can't be empty")
+    };
     gha_print(&json);
     gha_output("matrix", &json);
     Ok(())
```

Reply:
actually though, this point seems kinda moot, if it is empty then `targets.toml` is empty, which seems not likely

disregard this :)